### PR TITLE
#314 sign() and verifySignature() ribosome actions

### PR DESCRIPTION
**Pull request by jackThammer**

### CODE REVIEW

File: action.go [hunk 1]
```diff
@@ -327,6 +327,69 @@ func (a *ActionMakeHash) Do(h *Holochain) (response interface{}, err error) {
...
+	return "sign"
+}
+
+func (a *ActionSign) Args() []Arg {
+	return []Arg{{Name: "doc", Type: EntryArg}}
+}
+
+func (a *ActionSign) Do(h *Holochain) (response interface{}, err error) {
+	var b []byte
+	//hash, err = a.entry.Sum(h.hashSpec)
```
Comment (zippy) [inline action.go]: lets clean up this line before the final merge, I think it's left over from makeHash right?

File: holochain.go [hunk 2]
```diff
@@ -612,6 +614,33 @@ func (h *Holochain) Send(proto Protocol, to peer.ID, t MsgType, body interface{}
 	return
 }
 
+func (h *Holochain) Sign(doc []byte) (sig []byte, err error) {
```
Comment (zippy) [inline holochain.go]: needs a comment before it like:
//Sign uses the agent' private key to sign the contents of doc

File: holochain.go [hunk 3]
```diff
@@ -612,6 +614,33 @@ func (h *Holochain) Send(proto Protocol, to peer.ID, t MsgType, body interface{}
...
+	privKey := h.agent.PrivKey()
+	sig, err = privKey.Sign(doc)
+	if err != nil {
+		return
+	}
+	return
+}
+
+//TODO
```
Comment (zippy) [inline holochain.go]: fix comment

File: action.go [hunk 4]
```diff
@@ -327,6 +327,69 @@ func (a *ActionMakeHash) Do(h *Holochain) (response interface{}, err error) {
...
+	return "verifySignature"
+}
+
+func (a *ActionVerifySignature) Args() []Arg {
+	return []Arg{{Name: "signature", Type: StringArg}, {Name: "data", Type: StringArg}, {Name: "pubKey", Type: StringArg}}
+}
+
+func (a *ActionVerifySignature) Do(h *Holochain) (response bool, err error) {
+	var b bool
+	//hash, err = a.entry.Sum(h.hashSpec)
```
Comment (zippy) [inline action.go]: same here

File: holochain.go [hunk 5]
```diff
@@ -612,6 +614,33 @@ func (h *Holochain) Send(proto Protocol, to peer.ID, t MsgType, body interface{}
...
+func (h *Holochain) VerifySignature(signature string, data string, pubKey string) (matches bool, err error) {
+	var pubKeyIC ic.PubKey
+	var sig []byte
+	sig = b58.Decode(signature)
```
Comment (zippy) [inline holochain.go]: On thinking about this, I think the b58 encode decode shouldn't be in this function, it should be in the Action.Do() function.  I think signature should be of type []byte and pubKey should be be and actual unmarshaled pubKey, and the decoding is part of the action.

**Comment by jackThammer** [comment]
I have committed the changes that were suggested by @zippy

**Comment by zippy** [comment]
One more thing @jackThammer, before doing the pull request, its usually a good idea to merge master into your working branch, so that you've also been able to run you changes against the tests in the latest branch.

**Comment by lucksus** [comment]
I've merge master into here and will confirm this PR now.

**Pull request merged**